src/understand-tidb/implementation-of-typical-operators.md

```diff
@@ -1 +1,142 @@
 # Implementation of Typical Operators
+
+This section will introduce the implementation details of 3 typical TiDB operators: Sort, HashAgg and HashJoin.
+
+First of all, all the operator should implement 3 basic interface of `Executor`
+
+* [Open](https://github.com/pingcap/tidb/blob/v7.4.0/executor/internal/exec/executor.go#L50) initialize the operator, setup memory tracker/disk tracker and other meta info for current operator
+* [Next](https://github.com/pingcap/tidb/blob/v7.4.0/executor/internal/exec/executor.go#L51) each call of `Next` will return a chunk of data, return an empty chunk means the execution is done for current executor. Note `Next` is not thread safe, it is by design that `Next` is not called concurrently for all the operators 
+* [Close](https://github.com/pingcap/tidb/blob/v7.4.0/executor/internal/exec/executor.go#L52) is responsible for release all the resource hold by the executor
+
+### Sort
+
+Sort operator is used to arrange the result set of a query in a specific order. In TiDB the operator implement sort is [`SortExec`](https://github.com/pingcap/tidb/blob/v7.4.0/executor/sort.go#L36). The basic idea in `SortExec` is to read all the data from its child executor, then sort the whole data set. 
+
+In `Next`, it calls [`fetchRowChunks`](https://github.com/pingcap/tidb/blob/v7.4.0/executor/sort.go#L179)  to read all the data from its child executor. `fetchRowChunks` tries to save all the data in one [`SortedRowContainer`](https://github.com/pingcap/tidb/blob/v7.4.0/util/chunk/row_container.go#L460), the memory usage increases as the amount of input data increases. In order to control the memory usage, `SortExec` supports spill to disk. The details of spill to disk is hidden inside `SortedRowContainer`, each time when insert a chunk into current `SortedRowContainer` returns `ErrCannotAddBecauseSorted`, it means current `SortedRowContainer` is spilled, `SortExec` will generate a new `SortedRowContainer` and insert the chunk to the new one. When there is no data from its child executor, `SortExec`  will [sort](https://github.com/pingcap/tidb/blob/v7.4.0/executor/sort.go#L241) current `SortedRowContainer`.
+
+After `fetchRowChunks` finishes, `Next` will begin to generate sorted results, depends weather spill to disk is triggered or not, there is two ways to generate the final sorted results
+
+* Spill is not triggered: this is the simple case, when spill is not triggered, since the whole `SortedRowContainer` is sorted at the end of `fetchRowChunks`, in `Next`, it just calls [`GetSortedRowAndAlwaysAppendToChunk`](https://github.com/pingcap/tidb/blob/v7.4.0/executor/sort.go#L133) to retrieve the sorted data from `SortedRowContainer`
```

Comment:
```suggestion
* Spill not initiated: In this straightforward scenario, if there's no spill, since the entire `SortedRowContainer` gets sorted at the end of `fetchRowChunks`, during `Next`, it simply invokes [`GetSortedRowAndAlwaysAppendToChunk`](https://github.com/pingcap/tidb/blob/v7.4.0/executor/sort.go#L133) to fetch the sorted data from `SortedRowContainer`.
```